### PR TITLE
Adopt table scan methods for Scanner

### LINKED ISSUE #2075: Refactor Scanner to use table scan interface
Since PG12, tables can have different access methods (storage). However, our Scanner code still uses heap access, which means it can only scan heap tables. 

We should adopt the table scan API to make  Scanner agnostic to the underlying storage of a table.

### PR DESCRIPTION
**Pull request by erimatnor**
This change makes the Scanner code agnostic to the underlying storage
implementation of the tables it scans. This also fixes a bug that made
it impossible to use non-heap table access methods on a
hypertable. The bug existed because a check is made for existing data
before a table is made into a hypertable. And, since this check reads
data from the table using the Scanner, it must be able to read the
data irrespective of the underlying storage.

As a result of the more generic scan interface, resource management is
also improved by delivering tuples in reference-counted tuple table
slots. A backwards-compatibility layer is used for PG11, which maps
all table access functions to the heap equivalents.

A potential memory corruption issue is also fixed when scanning chunk constraints (separate commit).

Fixes #2075

### CODE REVIEW

File: src/chunk_index.c [hunk 1]
```diff
@@ -863,8 +891,11 @@ chunk_index_tuple_rename(TupleInfo *ti, void *data)
 	else
 		namestrcpy(&chunk_index->index_name, info->newname);
 
-	ts_catalog_update(ti->scanrel, tuple);
-	heap_freetuple(tuple);
+	ts_catalog_update(ti->scanrel, new_tuple);
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
```
Comment (k-rus) [inline src/chunk_index.c:898]: Can it be freed directly after copying to `new_tuple` or use `materialise = true` instead of `new_tuple`?

File: src/planner.c [hunk 2]
```diff
@@ -33,8 +33,8 @@
 #include <parser/analyze.h>
 
 #include <catalog/pg_constraint.h>
-#include "compat.h"
 #include "compat-msvc-exit.h"
+#include "compat.h"
```
Comment (k-rus) [inline src/planner.c:37]: Is there any particular reason for this reordering? I doubt `compat.h` depends on `compat-msvc-exit.h`. If not, then it breaks lexicographical order and is unnecessary change for the PR.

File: src/scan_iterator.h [hunk 3]
```diff
@@ -37,16 +37,27 @@ ts_scan_iterator_tuple_info(ScanIterator *iterator)
...
+	return iterator->tinfo->slot;
+}
+
 static inline HeapTuple
 ts_scan_iterator_tuple(ScanIterator *iterator)
 {
-	return iterator->tinfo->tuple;
+	return ExecFetchSlotHeapTuple(iterator->tinfo->slot, false, NULL);
+}
+static inline HeapTuple
```
Comment (k-rus) [inline src/scan_iterator.h:51]: New line is missing:
```suggestion

static inline HeapTuple
```

File: src/scan_iterator.h [hunk 4]
```diff
@@ -37,16 +37,27 @@ ts_scan_iterator_tuple_info(ScanIterator *iterator)
...
+ts_scan_iterator_slot(ScanIterator *iterator)
+{
+	return iterator->tinfo->slot;
+}
+
 static inline HeapTuple
 ts_scan_iterator_tuple(ScanIterator *iterator)
 {
-	return iterator->tinfo->tuple;
+	return ExecFetchSlotHeapTuple(iterator->tinfo->slot, false, NULL);
```
Comment (k-rus) [inline src/scan_iterator.h:49]: Since `ts_scan_iterator_slot` is defined, shouldn't it be used already here?
```suggestion
	return ExecFetchSlotHeapTuple(ts_scan_iterator_slot(iterator), false, NULL);
```

File: src/scan_iterator.h [hunk 5]
```diff
@@ -37,16 +37,27 @@ ts_scan_iterator_tuple_info(ScanIterator *iterator)
...
+}
+static inline HeapTuple
+ts_scan_iterator_materialized_tuple(ScanIterator *iterator)
+{
+	return ExecFetchSlotHeapTuple(iterator->tinfo->slot, true, NULL);
```
Comment (k-rus) [inline src/scan_iterator.h:54]:
```suggestion
	return ExecFetchSlotHeapTuple(ts_scan_iterator_slot(iterator), true, NULL);
```

File: src/scan_iterator.h [hunk 6]
```diff
@@ -37,16 +37,27 @@ ts_scan_iterator_tuple_info(ScanIterator *iterator)
 	return iterator->tinfo;
 }
 
+static inline TupleTableSlot *
+ts_scan_iterator_slot(ScanIterator *iterator)
+{
+	return iterator->tinfo->slot;
+}
```
Comment (k-rus) [inline src/scan_iterator.h:44]: I don't see this function is used anywhere. Should it be dropped or used?

File: src/scanner.c [hunk 7]
```diff
@@ -306,3 +337,13 @@ ts_scanner_scan_one(ScannerCtx *ctx, bool fail_if_not_found, char *item_type)
 			return false;
 	}
 }
+
+TSDLLEXPORT ItemPointer
```
Comment (k-rus) [inline src/scanner.c:341]: We usually don't include exporting details such as this macro into C files, but this file seems to be a complete exception :)

File: tsl/src/continuous_aggs/options.c [hunk 8]
```diff
@@ -173,17 +180,25 @@ update_materialized_only(ContinuousAgg *agg, bool materialized_only)
 		bool nulls[Natts_continuous_agg];
 		Datum values[Natts_continuous_agg];
 		bool repl[Natts_continuous_agg] = { false };
-		HeapTuple new;
+		bool should_free;
+		HeapTuple tuple = ExecFetchSlotHeapTuple(ti->slot, false, &should_free);
+		HeapTuple new_tuple;
+		TupleDesc tupdesc = ts_scan_iterator_tupledesc(&iterator);
```
Comment (k-rus) [inline tsl/src/continuous_aggs/options.c:186]: I've noticed that other code files use the code of `ts_scan_iterator_tupledesc` to get tuple description. It would be nice homogenise that, e.g., all relevant cases use `ts_scan_iterator_tupledesc`.

File: src/utils.h [hunk 9]
```diff
@@ -96,6 +98,8 @@ extern TSDLLEXPORT List *ts_get_reloptions(Oid relid);
 
 #define STRUCT_FROM_TUPLE(tuple, mctx, to_type, form_type)                                         \
 	(to_type *) ts_create_struct_from_tuple(tuple, mctx, sizeof(to_type), sizeof(form_type));
```
Comment (k-rus) [inline src/utils.h:100]: I don't think it used anywhere as it is replaced with `STRUCT_FROM_SLOT`. Can it be removed?

File: src/bgw_policy/chunk_stats.c [hunk 10]
```diff
@@ -154,17 +157,19 @@ static ScanTupleResult
...
 {
 	TimestampTz *updated_last_time_job_run = data;
-	HeapTuple tuple = heap_copytuple(ti->tuple);
-	BgwPolicyChunkStats *chunk_stats = STRUCT_FROM_TUPLE(ti->tuple,
-														 ti->mctx,
-														 BgwPolicyChunkStats,
-														 FormData_bgw_policy_chunk_stats);
+	bool should_free;
+	HeapTuple tuple = ExecFetchSlotHeapTuple(ti->slot, false, &should_free);
+	HeapTuple new_tuple = heap_copytuple(tuple);
```
Comment (gayyappan) [inline src/bgw_policy/chunk_stats.c:162]: I am puzzled by this change: What do we gain by introducing a call to ExecFetchSlotHeapTuple at line 161.? We make a copy of the tuple here.

File: src/bgw/job.c [hunk 11]
```diff
@@ -184,9 +184,11 @@ get_job_type_from_name(Name job_type_name)
 }
 
 static BgwJob *
-bgw_job_from_tuple(HeapTuple tuple, size_t alloc_size, MemoryContext mctx)
+bgw_job_from_tupleinfo(TupleInfo *ti, size_t alloc_size, MemoryContext mctx)
 {
 	BgwJob *job;
+	bool should_free;
+	HeapTuple tuple = ExecFetchSlotHeapTuple(ti->slot, false, &should_free);
```
Comment (gayyappan) [inline src/bgw/job.c:191]: How does the memory behave before/after this change? I see code where we copy the tuple contents into alloc'ed memory. Could you elaborate on how/why ExecFetchSlotHeapTuple helps here.

File: src/chunk.c [hunk 12]
```diff
@@ -2603,7 +2608,7 @@ chunk_tuple_delete(TupleInfo *ti, DropBehavior behavior, bool preserve_chunk_cat
 
 	if (!preserve_chunk_catalog_row)
 	{
-		ts_catalog_delete(ti->scanrel, ti->tuple);
+		ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
```
Comment (gayyappan) [inline src/chunk.c:2611]: I see some places where ts_catalog_delete has been replaced by ts_catalog_delete_tid. What is the reasoning behind the change? Aren't they equivalent in this case?

Comment (erimatnor) [reply src/chunk.c:2611]: The table API wraps tuples in tuple slots, so we don't have direct access to the tuple unless it is potentially re-materialized, which comes at a cast. FWIW, `ts_catalog_delete` takes a tuple, but internally just calls `ts_catalog_delete_tid`.

File: src/bgw/job_stat.c [hunk 13]
```diff
@@ -117,8 +116,10 @@ ts_bgw_job_stat_delete(int32 bgw_job_id)
 static ScanTupleResult
 bgw_job_stat_tuple_mark_start(TupleInfo *ti, void *const data)
 {
-	HeapTuple tuple = heap_copytuple(ti->tuple);
-	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+	bool should_free;
+	HeapTuple tuple = ExecFetchSlotHeapTuple(ti->slot, false, &should_free);
+	HeapTuple new_tuple = heap_copytuple(tuple);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/bgw/job_stat.c:121]: This was my first inclination as well, but the following comment on `ExecFetchSlotHeapTuple` made me opt out of this:

```
* NB: If materialize is true, modifications of the returned tuple are
 * allowed. But it depends on the type of the slot whether such modifications
 * will also affect the slot's contents. While that is not the nicest
 * behaviour, all such modifications are in the process of being removed.```

File: src/scanner.c (hunk 7 above)
Comment (erimatnor) [reply src/scanner.c:341]: Yes, this was also the reason I added it here. I figured a separate PR can clean this up if necessary

File: src/utils.h (hunk 9 above)
Comment (erimatnor) [reply src/utils.h:100]: I'll take a look

File: src/scanner.c [hunk 14]
```diff
@@ -23,45 +23,60 @@ enum ScannerType
...
 typedef struct Scanner
 {
-	Relation (*openheap)(InternalScannerCtx *ctx);
+	Relation (*openscan)(InternalScannerCtx *ctx);
 	ScanDesc (*beginscan)(InternalScannerCtx *ctx);
 	bool (*getnext)(InternalScannerCtx *ctx);
 	void (*endscan)(InternalScannerCtx *ctx);
-	void (*closeheap)(InternalScannerCtx *ctx);
+	void (*closescan)(InternalScannerCtx *ctx);
 } Scanner;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/scanner.c:31]: This matches the calls on the underlying scan API it wraps. 

endscan ends the scan, but doesn't close the relation. In other words, you can start another scan after endscan without having to reopen the relation.

File: src/bgw/job_stat.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/bgw/job_stat.c:121]: So it wasn't only me who didn't understand :)

File: src/planner.c (hunk 2 above)
Comment (k-rus) [reply src/planner.c]: Thank you for clarifying!
